### PR TITLE
Extend migration framework

### PR DESCRIPTION
**Pull request by rphmeier**
Simple pair migrations are moved to `SimpleMigration` trait.
Introduces `Migration` trait which takes an input database and writes whatever it wants into an output database.

Tests now create temporary databases to migrate.

### CODE REVIEW

File: util/src/migration/mod.rs [hunk 1]
```diff
@@ -15,28 +15,208 @@
...
+
+		for (key, value) in source.iter() {
+
+			if let Some((key, value)) = self.simple_migrate(key.to_vec(), value.to_vec()) {
+				batch.insert(key, value);
+			}
+
+			if batch.len() == config.batch_size {
+				try!(commit_batch(dest, &batch));
+			}
```
Comment (debris) [inline util/src/migration/mod.rs]: After commit batch should be cleared. `batch = BTreeMap::new();`

File: util/src/migration/tests.rs [hunk 2]
```diff
@@ -14,19 +14,49 @@
...
+		let transaction = DBTransaction::new();
+		for (k, v) in pairs {
+			transaction.put(&k, &v).expect("failed to add pair to transaction");
+		}
+
+		db.write(transaction).expect("failed to write db transaction");
+	}
+}
+
+// helper for verifying a a migrated database.
```
Comment (debris) [inline util/src/migration/tests.rs]: a a

**Comment by debris** [comment]
I've tried migrating blockchain from 23.05 (1.5m blocks). Everything works fine

**Comment by rphmeier** [comment]
Also works with my accountdb migration

**Comment by rphmeier** [comment]
appveyor needs restarting, failure unrelated to changes

File: util/src/migration/mod.rs [hunk 3]
```diff
@@ -15,28 +15,211 @@
...
 }
+
+/// Get the path where all databases reside.
+fn database_path(path: &Path) -> PathBuf {
+	let mut temp_path = path.to_owned();
+	temp_path.pop();
+	temp_path
+}
+
+enum TempIdx {
```
Comment (gavofyork) [inline util/src/migration/mod.rs]: `TempIndex`: don't kill the vowels!

File: util/src/migration/tests.rs [hunk 4]
```diff
@@ -14,19 +14,50 @@
...
+
+fn db_path(path: &Path) -> PathBuf {
+	let mut p = path.to_owned();
+	p.push("db");
+	p
+}
+
+// initialize a database at the given directory with the given values.
+fn make_db(path: &Path, pairs: BTreeMap<Vec<u8>, Vec<u8>>) {
+	let db = Database::open_default(path.to_str().unwrap()).expect("failed to open temp database");
```
Comment (gavofyork) [inline util/src/migration/tests.rs]: PROVE OR REMOVE.

File: util/src/migration/tests.rs [hunk 5]
```diff
@@ -14,19 +14,50 @@
...
+	{
+		let transaction = DBTransaction::new();
+		for (k, v) in pairs {
+			transaction.put(&k, &v).expect("failed to add pair to transaction");
```
Comment (gavofyork) [inline util/src/migration/tests.rs]: PROVE OR REMOVE.

File: util/src/migration/tests.rs [hunk 6]
```diff
@@ -14,19 +14,50 @@
...
+		}
+
+		db.write(transaction).expect("failed to write db transaction");
```
Comment (gavofyork) [inline util/src/migration/tests.rs]: PROVE OR REMOVE.

File: util/src/migration/tests.rs [hunk 7]
```diff
@@ -14,19 +14,50 @@
...
+	}
+}
+
+// helper for verifying a migrated database.
+fn verify_migration(path: &Path, pairs: BTreeMap<Vec<u8>, Vec<u8>>) {
+	let db = Database::open_default(path.to_str().unwrap()).unwrap();
+
+	for (k, v) in pairs {
+		let x = db.get(&k).unwrap().unwrap();
```
Comment (gavofyork) [inline util/src/migration/tests.rs]: PROVE OR REMOVE.

File: parity/migration.rs [hunk 8]
```diff
@@ -65,6 +64,16 @@ impl From<IoError> for Error {
 	}
 }
 
+impl From<MigrationError> for Error {
+	fn from(err: MigrationError) -> Self {
+		match err {
+			MigrationError::Io(e) => Error::Io(e),
+			_ => Error::MigrationFailed,
+
```
Comment (gavofyork) [inline parity/migration.rs]: superfluous newline.

**Comment by gavofyork** [comment]
some panickers introduced in there and style issues. looks ok otherwise.

File: util/src/migration/tests.rs (hunk 4 above)
Comment (debris) [inline util/src/migration/tests.rs]: I believe these unwraps are harmless cause these are helper functions for test code.

File: util/src/migration/tests.rs (hunk 5 above)
Comment (rphmeier) [inline util/src/migration/tests.rs]: test code -- fail hard and fail fast

File: util/src/migration/tests.rs (hunk 6 above)
Comment (rphmeier) [inline util/src/migration/tests.rs]: test code -- fail hard and fail fast

File: util/src/migration/tests.rs (hunk 7 above)
Comment (rphmeier) [inline util/src/migration/tests.rs]: test code -- fail hard and fail fast

**Pull request merged**